File: src/infrastructure/rest/contracts/diagrams.py
"""Response contracts for the diagram surface.

The list-shaped reads answer an object with ``items`` rather than a bare array. A top-level array
has nowhere to put a total or a cursor, so adding either later would be a second breaking change —
and an array also cannot be a named component in the generated document, which is what the
frontend's type generation keys off.

Diagram *content* stays open where it belongs to a diagram-type module: a datatype classifier's
attributes, an activity step's shape, an entity's display block. Those are the module's vocabulary,
not this contract's, and closing them here would put the delivery layer in the business of every
diagram kind. Everything the surface itself decides is declared.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict

from src.infrastructure.rest.contracts.connections import ConnectionSummary
from src.infrastructure.rest.contracts.entities import (
    ContextConnection,
    EntityDisplayItemResponse,
    EntitySummary,
)
from src.infrastructure.rest.contracts.viewpoints import ViewpointApplicationResponse
from src.infrastructure.rest.contracts.wire_nulls import NullsOmitted
from src.infrastructure.rest.contracts.wire_shape import Closed


class _ModuleShaped(BaseModel):
    """A payload whose fields belong to a diagram-type module rather than to this surface.

    Open on purpose, and named so the exception is auditable: the alternative is a delivery-layer
    model that has to change whenever a diagram kind adds a field, which would make the ontology's
    extensibility depend on this file.
    """

    model_config = ConfigDict(extra="allow")


class DiagramReference(Closed):
    """A diagram that draws a given source/target pair."""

    artifact_id: str
    name: str


class DiagramReferenceListResponse(Closed):
    """Which diagrams draw a given pair — what a rename or removal has to warn about."""

    items: list[DiagramReference]


class DiagramTypeMemberItem(_ModuleShaped):
    """One entity or connection type a diagram kind accepts, with its presentation data."""

    key: str | None = None
    label: str | None = None


class DiagramTypeEntityTypeListResponse(Closed):
    items: list[DiagramTypeMemberItem]


class DiagramTypeConnectionTypeListResponse(Closed):
    items: list[DiagramTypeMemberItem]


class DiagramSummary(Closed):
    """One row of the diagram list.

    Four fields left with this release — ``keywords``, ``host_diagram_id``, ``tlp`` and
    ``viewpoint``. ``diagram_to_summary`` never filled any of them, so every row carried an empty
    list and three nulls, and the client stripped all four on decode. A field the producer does not
    fill is not a smaller promise than a wrong one; the pinned viewpoint is reported by the diagram
    read, which does resolve it.
    """

    artifact_id: str
    diagram_type: str
    name: str
    version: str
    status: str
    path: str
    is_global: bool
    group: str | None = None
    last_updated: str | None = None


class DiagramListResponse(Closed):
    """A page of diagrams, with the count of the *filtered* population."""

    total: int
    items: list[DiagramSummary]


class DatatypeClassifierInfo(Closed):
    """One classifier a datatype diagram declares, as the picker needs it.

    ``host_diagram_id`` is the diagram that owns it — a classifier is a diagram-local construct, so
    it has no file of its own and its full identifier is composite (see ``_diagram_entity_extraction``).
    Empty for a workspace-scoped type, which is the difference ``scope`` names.
    """

    type_id: str
    label: str
    kind: str
    scope: str
    host_diagram_id: str


class DatatypeTypeListResponse(Closed):
    """A page of datatype classifier types, with the primitives they may be built from.

    ``generation`` is the index generation the page was read at: a picker holding a page and then
    resolving one of its entries needs to know the two came from the same model, and a stale page that
    silently resolves against a newer index is how a deleted classifier gets offered.

    ``next_cursor`` is null on the last page, following the house pagination convention — present and
    null rather than absent, so a client can tell "no more" from "this server does not paginate".
    """

    generation: int
    primitives: list[str]
    classifiers: list[DatatypeClassifierInfo]
    next_cursor: str | None


class DatatypeTypeUsage(Closed):
    """One place a classifier type is referenced as an attribute's type."""

    diagram_id: str
    classifier_local_id: str
    attr_name: str


class DatatypeTypeUsageResponse(Closed):
    """Everywhere one classifier type is used.

    ``type_id`` is echoed although the caller supplied it in the path: a client that fans out over
    several types and collects the answers needs each one to say which question it answers, and
    correlating by request order is the kind of thing that works until it does not.
    """

    type_id: str
    usages: list[DatatypeTypeUsage]


class MatrixConnTypeConfig(Closed):
    """One relationship type the matrix draws, and whether it is currently shown."""

    conn_type: str
    active: bool


class MatrixPreviewResponse(Closed):
    """A matrix write's dry run: the rendered body it would store, and nothing else.

    The route declared ``WriteResultResponse`` — the six-key mutation envelope — while returning
    ``{"markdown": …}``, so FastAPI's response validation raised on every single call and the
    preview answered 500 with a body that deliberately carries no diagnostic. Nothing noticed
    because a preview is a write-shaped operation, and ``NEVER_REQUESTED_OPERATIONS`` recorded
    ``matrices_preview_matrix`` as never once having answered 2xx: the Preview button on both matrix
    views has never worked through the running server.

    A dry run is not a mutation and does not share its envelope. There is no ``wrote``, no ``path``
    and no ``artifact_id``, because nothing was written and nothing has an address yet.
    """

    #: The matrix body the write would store, with entity ids already linkified.
    markdown: str


class MatrixConfigResponse(Closed):
    """A matrix diagram's authored configuration, parsed out of its PUML frontmatter.

    ``from_entity_ids`` and ``to_entity_ids`` are **present and null** for a square matrix rather than
    absent: null says "no separate axis was authored, so both axes are ``entity_ids``", which is a
    different statement from a field that was never part of the response. The producer emits the keys
    unconditionally, so the contract does too.
    """

    artifact_id: str
    name: str
    status: str
    version: str
    keywords: list[str]
    #: The population when both axes are the same — the square case.
    entity_ids: list[str]
    #: The row axis of an asymmetric matrix; null when the matrix is square.
    from_entity_ids: list[str] | None
    #: The column axis of an asymmetric matrix; null when the matrix is square.
    to_entity_ids: list[str] | None
    conn_type_configs: list[MatrixConnTypeConfig]
    #: Whether the cells combine every active relationship type into one mark.
    combined: bool
    #: The diagram's authored PUML body, which the matrix owns rather than deriving.
    matrix_body: str


class EntityDisplaySearchResponse(Closed):
    """A page of entities a user could place on a diagram.

    ``next_cursor`` is an offset the caller passes back verbatim; null means this page is the last.
    The cursor is opaque by contract — the ordering it encodes is the server's to change.
    """

    items: list[EntityDisplayItemResponse]
    next_cursor: str | None


class HopSuggestionGroup(Closed):
    """Entities one hop further out than the last group, so a picker can offer "and their
    neighbours" without asking the user to think in graph distance.

    ``hop`` starts at 1: hop 0 is what the diagram already holds, and repeating it would invite a
    surface to offer removing an entity as a suggestion to add one.
    """

    hop: int
    items: list[EntityDisplayItemResponse]


class DiagramEntityDiscoveryResponse(Closed):
    """Three ways to find the next thing to put on a diagram, answered together.

    A search over the whole model, the connections that would appear if the entities already
    selected were placed, and what lies one, two or three hops out. One request rather than three
    because the panel shows all three at once, and three would race.
    """

    search_results: list[EntityDisplayItemResponse]
    #: Connections between the entities already selected — what the diagram would gain, not what it
    #: has. Excluded from ``search_results`` by construction: an entity already placed is not a
    #: candidate to place.
    candidate_connections: list[ContextConnection]
    suggested_entities: list[HopSuggestionGroup]


class DerivedViewEntityResponse(Closed):
    """One entity a model-backed diagram derived rather than the author placing it.

    ``item_type`` and ``role`` are the diagram-type module's own vocabulary — generic code forwards
    them and never interprets them. ``role`` is load-bearing all the same: the scope root arrives as
    ``scope`` and cannot be excluded, so a checklist offering to remove it offers something the
    engine will not do.
    """

    id: str
    name: str
    item_type: str
    role: str
    excluded: bool


class DiagramPreviewResponse(Closed):
    """What a diagram write would produce, without writing it.

    ``image`` is null when rendering failed and ``warnings`` says why — a preview that cannot render
    is still a successful answer about the write. ``derived_entities`` is null when the diagram type
    derives nothing, which is not the same as deriving nothing *this time*: an empty list means the
    projection ran and found no candidates, and the two read differently to a user.
    """

    puml: str
    image: str | None
    warnings: list[str]
    derived_entities: list[DerivedViewEntityResponse] | None


class DiagramDetailResponse(NullsOmitted):
    """One diagram, read whole: the record, its source, and what the source declares.

    ``entity_ids_used``/``connection_ids_used`` are present only when the file's frontmatter
    declares them — an older diagram simply does not, and an empty list would claim it draws
    nothing.

    ``type_extras`` is where a diagram-type module's own additions arrive. They used to be merged
    into this envelope, which made the whole response an open object promising nothing — a client
    reading the schema was told a diagram read returns "these fields, and possibly anything". One
    declared field carries the module's territory instead, and the envelope closes.
    """

    artifact_id: str
    artifact_type: str
    record_type: Literal["diagram"]
    name: str
    diagram_type: str
    version: str
    status: str
    path: str
    is_global: bool
    group: str | None = None
    last_updated: str | None = None
    content_snippet: str
    puml_source: str
    #: The rendered image beside the source file, absent when nothing has been rendered yet.
    rendered_filename: str | None = None
    entity_ids_used: list[str] | None = None
    connection_ids_used: list[str] | None = None
    #: The diagram-kind's own placement data, keyed by entity id. Absent when the diagram places
    #: nothing of its own — a hand-drawn ArchiMate view, say.
    diagram_entities: dict[str, Any] | None = None
    #: Frontmatter this surface does not model, returned as written.
    extra: dict[str, Any] | None = None
    viewpoint: ViewpointApplicationResponse | None = None
    #: This diagram kind's own additions — a matrix's rendered body, and whatever a future kind
    #: contributes. Absent for a kind that adds nothing.
    type_extras: dict[str, Any] | None = None


class DiagramContextEntity(EntitySummary):
    """One entity as this diagram places it: the list row, plus the alias it is drawn under.

    ``display_alias`` is here and not on :class:`EntitySummary` because only a diagram read resolves
    it — the entity list has no diagram to draw the entity on, and a field the list route never
    fills would read as "this entity has no alias".
    """

    display_alias: str


class DiagramContextConnection(ConnectionSummary, NullsOmitted):
    """One connection as this diagram draws it: the record, plus how it is rendered here.

    The aliases and ``edge_key`` are the PlantUML identity of the edge in *this* file, which is what
    an overlay needs to find the drawn line for a connection. ``edge_label_override`` is the label
    the author set for that edge and belongs to the diagram, not to the connection — the same
    connection drawn on two diagrams can carry two labels.
    """

    source_alias: str
    target_alias: str
    edge_key: str
    edge_label_override: str | None = None


class DiagramEntityListResponse(NullsOmitted):
    """Entities placed on one diagram.

    The rows are :class:`DiagramContextEntity` because they are *the same rows* — the list route and
    the context read both return what ``diagram_entities_and_puml`` produced, and the connection pair
    below does the same for ``diagram_context_payload``'s ``connections``. They used to be declared
    as a pair of open three-field placeholders, which is how one producer came to have two contracts:
    the context read declared the real shape and omitted its unset optionals, while these two
    published a shape nothing produced and sent `"last_updated": null` on the wire. The client
    decodes both with the context read's schema, so the entity list threw in the decoder before a
    row was drawn — the FMEA defect's exact shape, one release later, on a route no browser spec
    happens to read.
    """

    items: list[DiagramContextEntity]


class DiagramConnectionListResponse(NullsOmitted):
    """Connections drawn on one diagram — literally ``diagram_context_payload``'s ``connections``."""

    items: list[DiagramContextConnection]


class DiagramContextResponse(NullsOmitted):
    """A diagram and everything an editor needs to work on it, in one read.

    Assembled server-side because an editor that fetched these separately could render a diagram
    against one model generation and its candidate connections against another. ``generation`` and
    ``etag`` are in the body for exactly that reason: they say which snapshot all of it came from.

    ``explicit_connection_pairs`` are the source/target alias pairs the PUML actually draws, which
    is how a stated connection is told from one that merely exists between two placed entities.
    """

    diagram: DiagramDetailResponse
    entities: list[DiagramContextEntity]
    connections: list[DiagramContextConnection]
    candidate_connections: list[ContextConnection]
    suggested_entities: list[HopSuggestionGroup]
    explicit_connection_pairs: list[tuple[str, str]]
    generation: int
    etag: str
    #: This diagram kind's own additions — a C4 diagram's navigation, and whatever a future kind
    #: contributes. Absent for a kind that adds nothing.
    type_extras: dict[str, Any] | None = None

File: tools/gui/src/domain/schemas/diagrams.ts
import { Schema } from 'effect'
import { ViewpointApplicationSchema } from './viewpoints'
import { EntityContextConnectionSchema } from './connections'
import { EntityDisplayInfoSchema } from './entities'

/**
 * One labelled box a diagram draws that the model does not hold.
 *
 * Recursive, because boxes nest. `entity-ids` keeps its frontmatter spelling on the wire, so what is
 * read is what is written back, and a member may name an entity or a single occurrence of one.
 */
export interface AuthoredGroupingWire {
  readonly label: string
  readonly 'entity-ids': readonly string[]
  readonly groups?: readonly AuthoredGroupingWire[]
  /** An override the backend still honours; the look is otherwise derived from the members. */
  readonly stereotype?: string
}

export const AuthoredGroupingSchema: Schema.Schema<AuthoredGroupingWire> = Schema.Struct({
  label: Schema.String,
  'entity-ids': Schema.Array(Schema.String),
  groups: Schema.optional(Schema.Array(Schema.suspend(() => AuthoredGroupingSchema))),
  stereotype: Schema.optional(Schema.String),
})

export const DiagramDetailSchema = Schema.Struct({
  artifact_id: Schema.String,
  artifact_type: Schema.String,
  name: Schema.String,
  diagram_type: Schema.String,
  version: Schema.String,
  status: Schema.String,
  record_type: Schema.Literal('diagram'),
  path: Schema.String,
  is_global: Schema.Boolean,
  group: Schema.optional(Schema.String),
  last_updated: Schema.optional(Schema.String),
  content_snippet: Schema.String,
  puml_source: Schema.String,
  /** Absent when nothing has been rendered yet. */
  rendered_filename: Schema.optional(Schema.String),
  entity_ids_used: Schema.optional(Schema.Array(Schema.String)),
  connection_ids_used: Schema.optional(Schema.Array(Schema.String)),
  diagram_entities: Schema.optional(Schema.Record({ key: Schema.String, value: Schema.Unknown })),
  /** The labelled boxes the diagram draws that the model does not hold. Read as well as written,
   * because an editor that could author one without seeing the existing ones would replace them. */
  authored_groupings: Schema.optional(Schema.Array(AuthoredGroupingSchema)),
  extra: Schema.optional(Schema.Record({ key: Schema.String, value: Schema.Unknown })),
  /** This diagram kind's own additions — a matrix's rendered body, and whatever a future kind
   * contributes. Absent for a kind that adds nothing. */
  type_extras: Schema.optional(Schema.Record({ key: Schema.String, value: Schema.Unknown })),
  viewpoint: Schema.optional(ViewpointApplicationSchema),
})
export type DiagramDetail = typeof DiagramDetailSchema.Type

// ── Matrix diagram ───────────────────────────────────────────────────────────

export const MatrixConnTypeConfigSchema = Schema.Struct({
  conn_type: Schema.String,
  active: Schema.Boolean,
})
export type MatrixConnTypeConfig = typeof MatrixConnTypeConfigSchema.Type

export const MatrixConfigSchema = Schema.Struct({
  artifact_id: Schema.String,
  name: Schema.String,
  status: Schema.String,
  version: Schema.String,
  keywords: Schema.Array(Schema.String),
  entity_ids: Schema.Array(Schema.String),
  // Present and null for a square matrix, never absent: the route emits both keys unconditionally, and
  // null says "no separate axis was authored" — a statement the absent case cannot make.
  from_entity_ids: Schema.NullOr(Schema.Array(Schema.String)),
  to_entity_ids: Schema.NullOr(Schema.Array(Schema.String)),
  conn_type_configs: Schema.Array(MatrixConnTypeConfigSchema),
  combined: Schema.Boolean,
  matrix_body: Schema.String,
})
export type MatrixConfig = typeof MatrixConfigSchema.Type

export const MatrixPreviewResultSchema = Schema.Struct({
  markdown: Schema.String,
})
export type MatrixPreviewResult = typeof MatrixPreviewResultSchema.Type

export const C4NavLinkSchema = Schema.Struct({
  diagram_id: Schema.String,
  diagram_name: Schema.String,
  diagram_type: Schema.String,
  scope_entity_id: Schema.optional(Schema.NullOr(Schema.String)),
})
export type C4NavLink = typeof C4NavLinkSchema.Type

export const C4NavigationSchema = Schema.Struct({
  current_level: Schema.Number,
  scope_entity_id: Schema.NullOr(Schema.String),
  scope_entity_name: Schema.NullOr(Schema.String),
  parent_diagrams: Schema.Array(C4NavLinkSchema),
  child_diagrams: Schema.Array(C4NavLinkSchema),
})
export type C4Navigation = typeof C4NavigationSchema.Type

/** One entity as this diagram places it: the list row, plus the alias it is drawn under.
 *
 * A struct of its own rather than `EntitySummarySchema`, which is the *list* row and carries
 * several optional fields no diagram read fills. `display_alias` is required here because only a
 * diagram read resolves it. */
export const DiagramContextEntitySchema = Schema.Struct({
  artifact_id: Schema.String,
  artifact_type: Schema.String,
  name: Schema.String,
  version: Schema.String,
  status: Schema.String,
  domain: Schema.String,
  subdomain: Schema.String,
  path: Schema.String,
  is_global: Schema.Boolean,
  display_alias: Schema.String,
  group: Schema.optional(Schema.String),
  specializations: Schema.Array(Schema.String),
  host_diagram_id: Schema.optional(Schema.String),
  conn_in: Schema.optional(Schema.Number),
  conn_sym: Schema.optional(Schema.Number),
  conn_out: Schema.optional(Schema.Number),
  last_updated: Schema.optional(Schema.String),
})
export type DiagramContextEntity = typeof DiagramContextEntitySchema.Type

/** One connection as this diagram draws it: the record, plus its rendered identity here.
 *
 * The aliases and `edge_key` locate the drawn line in *this* file, and
 * `edge_label_override` is the label the author set for that edge — it belongs to the diagram, not
 * to the connection, since the same connection drawn twice can carry two labels. */
export const DiagramContextConnectionSchema = Schema.Struct({
  artifact_id: Schema.String,
  source: Schema.String,
  target: Schema.String,
  conn_type: Schema.String,
  version: Schema.String,
  status: Schema.String,
  path: Schema.String,
  content_text: Schema.String,
  associated_entities: Schema.Array(Schema.String),
  source_name: Schema.String,
  target_name: Schema.String,
  source_alias: Schema.String,
  target_alias: Schema.String,
  edge_key: Schema.String,
  edge_label_override: Schema.optional(Schema.String),
  src_multiplicity: Schema.optional(Schema.String),
  tgt_multiplicity: Schema.optional(Schema.String),
  specializations: Schema.Array(Schema.String),
  metadata: Schema.Record({ key: Schema.String, value: Schema.Unknown }),
  gar_artifact_id: Schema.optional(Schema.String),
})
export type DiagramContextConnection = typeof DiagramContextConnectionSchema.Type

/** Entities one hop further out than the last group. `hop` starts at 1 — hop 0 is what the
 * diagram already holds. */
export const HopSuggestionGroupSchema = Schema.Struct({
  hop: Schema.Number,
  items: Schema.Array(EntityDisplayInfoSchema),
})

export const DiagramContextSchema = Schema.Struct({
  diagram: DiagramDetailSchema,
  entities: Schema.Array(DiagramContextEntitySchema),
  connections: Schema.Array(DiagramContextConnectionSchema),
  candidate_connections: Schema.Array(EntityContextConnectionSchema),
  suggested_entities: Schema.Array(HopSuggestionGroupSchema),
  /** Source/target alias pairs the PUML actually draws — how a stated connection is told from
   * one that merely exists between two placed entities. */
  explicit_connection_pairs: Schema.Array(Schema.Tuple(Schema.String, Schema.String)),
  generation: Schema.Number,
  etag: Schema.String,
  /** This diagram kind's own additions — a C4 diagram's navigation, and whatever a future kind
   * contributes. Absent for a kind that adds nothing. */
  type_extras: Schema.optional(Schema.Record({ key: Schema.String, value: Schema.Unknown })),
})
export type DiagramContext = typeof DiagramContextSchema.Type

export const DiagramEntityDiscoverySchema = Schema.Struct({
  search_results: Schema.Array(EntityDisplayInfoSchema),
  candidate_connections: Schema.Array(EntityContextConnectionSchema),
  suggested_entities: Schema.Array(HopSuggestionGroupSchema),
})
export type DiagramEntityDiscovery = typeof DiagramEntityDiscoverySchema.Type

/** One entity a model-backed diagram derived rather than the author placing it.
 *
 * `item_type` and `role` are the diagram-type module's vocabulary and are forwarded, never
 * interpreted. `role` is load-bearing all the same: the scope root arrives as `scope` and the
 * engine will not exclude it, so a checklist offering to uncheck it offers nothing. Both fields
 * were absent here and stripped on decode, which is why that has gone unnoticed. */
export const DerivedEntitySchema = Schema.Struct({
  id: Schema.String,
  name: Schema.String,
  item_type: Schema.String,
  role: Schema.String,
  excluded: Schema.Boolean,
})
export type DerivedEntity = typeof DerivedEntitySchema.Type

export const DiagramPreviewResultSchema = Schema.Struct({
  puml: Schema.String,
  image: Schema.NullOr(Schema.String),
  warnings: Schema.Array(Schema.String),
  derived_entities: Schema.NullOr(Schema.Array(DerivedEntitySchema)),
})
export type DiagramPreviewResult = typeof DiagramPreviewResultSchema.Type
